### ISSUE TITLE
Web Extension fails to load on VS Code desktop after first install (works fine on VS Code for Web or after restarting)

### ISSUE THREAD

**Issue by billti** [microsoft/vscode#173779]
Does this issue occur when all extensions are disabled?: It is an extension loading issue

- VS Code Version: 1.75.0
- OS Version: macOS Ventura 13.2

Steps to Reproduce:

- Open an empty folder 
- Create a file `sample.qsx` and open the file and pin it in the editor
- Install `billti.qsx` extension

🐛 Error activating `billti.qsx` extension

<img width="484" alt="image" src="https://user-images.githubusercontent.com/10746682/222399866-e095f0f8-3083-44e3-935f-91c37dfdf9a7.png">

<details>
  <summary>Description From Author</summary>

I am building an extension as a pure web extension. It runs great in all scenarios (on github.dev, locally installed, under the debugger), except that when first installed on VS Code for Desktop it fails to load the resources from the extension package. Once I restart VS Code it starts working fine. (It works straight away in VS Code for Web).

1. Clone the very small repo at <https://github.com/billti/qsx-demo> to ease the repro (as it has a file of the correct type and the extension is recommended).
2. Open VS Code locally at the root of the repo, and install the recommended extension (`billti.qsx`)
3. Open the QSX file and see it is mostly working (highlighting, completions, errors). There are two commands the extension installs, `Show QSX panel` and `Debug QSX file`. Neither will work as intended, and the below errors will be shown in the console of the DevTools. (Note the first line error about the DebugAdapterDescriptorFactory failing, and the two GETs which result in a 401 error).

```text
workbench.desktop.main.js:2020 Activating extension 'billti.qsx' failed: a DebugAdapterDescriptorFactory can only be registered from the extension that defines the 'qsx' debugger..

index.html?id=17feadf8-9352-426a-b23d-567d72a6ad53&origin=49c6e153-5802-4051-be31-fc9fb031dfdd&swVersion=4&extensionId=nullExtensionDescription&platform=electron&vscode-resource-base-authority=vscode-resource.vscode-cdn.net&parentOrigin=vscode-file%3A%2F%2Fvscode-app:973          

GET https://file+.vscode-resource.vscode-cdn.net/Users/billti/.vscode/extensions/billti.qsx-0.1.5/dist/webview-ui-toolkit/toolkit.min.js net::ERR_ABORTED 401

(anonymous) @ index.html?id=17feadf8-9352-426a-b23d-567d72a6ad53&origin=49c6e153-5802-4051-be31-fc9fb031dfdd&swVersion=4&extensionId=nullExtensionDescription&platform=electron&vscode-resource-base-authority=vscode-resource.vscode-cdn.net&parentOrigin=vscode-file%3A%2F%2Fvscode-app:973
setTimeout (async)
onFrameLoaded @ index.html?id=17feadf8-9352-426a-b23d-567d72a6ad53&origin=49c6e153-5802-4051-be31-fc9fb031dfdd&swVersion=4&extensionId=nullExtensionDescription&platform=electron&vscode-resource-base-authority=vscode-resource.vscode-cdn.net&parentOrigin=vscode-file%3A%2F%2Fvscode-app:971
(anonymous) @ index.html?id=17feadf8-9352-426a-b23d-567d72a6ad53&origin=49c6e153-5802-4051-be31-fc9fb031dfdd&swVersion=4&extensionId=nullExtensionDescription&platform=electron&vscode-resource-base-authority=vscode-resource.vscode-cdn.net&parentOrigin=vscode-file%3A%2F%2Fvscode-app:1002

GET https://file+.vscode-resource.vscode-cdn.net/Users/billti/.vscode/extensions/billti.qsx-0.1.5/dist/webview.js net::ERR_ABORTED 401
```

5. Restart VS Code and try again, and notice that the extension and its WebView loads and works fine (i.e. the same two commands in the command palette work as expected - the QSX panel shows the table as expected, and the (mock) debugger starts successfully. DevTools shows no error regarding `DebugAdapterDescriptorFactory`, and the same URIs are being requested as after initial install when they failed, but this time they load successfully (the bottom two requests below).

<img width="997" alt="image" src="https://user-images.githubusercontent.com/993909/217476689-faeb15bb-36c8-4844-b389-125a6fe3b282.png">

Interestingly, if you open the exact same repo in VS Code for the Web via the browser at <https://github.dev/billti/qsx-demo>, and install the exact same extension, everything works straight away with no need to reload the browser.

Happy to provide code snippets on how I'm building Uris and attempting to fetch them if of use.

</details>

**Comment by billti** [comment]
Hi @alexdima . I'm a Microsoft employee (working on an effort to re-write the Q# developer kit tooling - https://azure.microsoft.com/en-us/resources/development-kit/quantum-computing/). If you want access to any internal repos, let me know (or I can make something minimal public, but will take some time to refactor for what I currently have). I can also just on a quick teams call if you do want to investigate and that would help. Thanks!

**Comment by billti** [comment]
OK. I think I've debugged this to the root cause, but I don't know why it's occurring. (Or thus how to fix it).

It looks like when the extension is first loaded the extension script itself tries to execute (obviously), and one of the actions it does is the `require('vscode')` call to get the API. However in the resulting `load` call at https://github.com/microsoft/vscode/blob/main/src/vs/workbench/api/common/extHostRequireInterceptor.ts#L164 I can see when it fails (again, if using the extension after install without a restart) the statement `const ext = this._extensionPaths.findSubstr(parent);` is returning undefined, so it falls through to return `this._defaultApiImpl = this._apiFactory(nullExtensionDescription, this._extensionRegistry, this._configProvider);`.

After VS Code is restarted, the extensionPath for the recently installed extension is present, and the calls it makes to `require('vscode')` result in success in the `load` method pointed to above.

In the failure case when it returns the `nullExtensionDescription`, then when the extension script's `activate` subsequently calls `vscode.debug.registerDebugAdapterDescriptorFactory('qsx', factory)`,  then as the nullExtension doesn't implement the debugger specified an exception is thrown at `https://github.com/microsoft/vscode/blob/main/src/vs/workbench/api/common/extHostDebugService.ts#L330`

I'll try and find some more time to investigate later in the week. If that gives you any clues to the cause/fix, then I'm happy to try and implement a fix with some guidance and open a PR. Thanks.

**Comment by billti** [comment]
PR up. That took a looooooot of debugging to get to the bottom of. Happy to update based on feedback. Thanks!

**Comment by sandy081** [comment]
@billti Thanks for detailed description and the PR. I have investigated and shortened the steps to reproduce.

@alexdima Here is my investigation details:

It seems web worker extension host is set to start lazily. So when there are no web extensions we do not star them. But we start it as soon as a web extension is installed like in above scenario. But unfortunately, the web extension host is using cached scanned extensions which does not have the installed web extension. Therefore it is failing to activate the extension. 

Hence moving this to you. Please let me know if there is anything I can help here.

**Comment by billti** [comment]
Thanks. That matches my finding and summarizes what I wrote in the PR (#174175 )

As outlined there, the simplest fix (and what seemed to be correct regardless of the lazy loading cause) was to invalid the cached list of scanned extension upon changing the installed extensions, and then repopulated it when next requested, which is what I implemented, and works in my testing. Happy to update if that's not the right appoach.

Thanks!
